### ISSUE TITLE
line-gradient colours are inconsistent when zooming into a geojson line

### ISSUE THREAD

**Issue by LR-brendan** [mapbox/mapbox-gl-js#8969]
I noticed as you zoom into a single geojson line with `line-gradient` styling that the style seems to 'reset'. 
I first noticed this in my application, but it is also present on the examples page on the Mapbox site with all browsers I've tried:
![image](https://i.imgur.com/xH6BMqj.gif)
The above gif is just an example, but this behaviour occurs all along the line. e.g zoom into the bridge.

**mapbox-gl-js version**: v1.5.0

**browsers**: 
Chrome 73.0.3683.86 
Firefox 69.03
Edge 41.16299.1419.0

### Steps to Trigger Behavior

 1. Zoom into line > level 14

### Link to Demonstration
https://docs.mapbox.com/mapbox-gl-js/example/line-gradient/
https://output.jsbin.com/kesogobuvu (as above, but with buttons to zoom to problem areas)

### Expected Behavior
Line-gradient colours should be consistent across all zoom levels for a single line.

### Actual Behavior
Colour gradient gets reset at multiple points along line as you zoom in.

**Comment by ryanhamley** [comment]
Thanks for the thorough report @LR-brendan! I was able to reproduce the problem in your examples. On a hunch, in JSBin I set `map.showTileBoundaries = true` and this is happening at tile boundaries as shown below.

<img width="636" alt="Screen Shot 2019-11-12 at 11 38 34 AM" src="https://user-images.githubusercontent.com/4523080/68704672-858f5e00-0541-11ea-8dd7-f0909d1f235e.png">

@mourner This is a regression which I tracked to https://github.com/mapbox/mapbox-gl-js/pull/8303 where the way we calculate line distance is changed.
